Block Bindings: Don't show protected fields that are bound to blocks and post meta

SantosGuillamot opened the pull request:
As [discussed in Gutenberg plugin](https://github.com/WordPress/gutenberg/pull/59326), this pull request adds two safety checks to ensure block bindings don't leak private post meta:

* Check if the meta field is protected.
* Check if the meta field is available through the REST API.

It seems safer to add these limitations to ensure no unwanted data is leaked. And it can be explored later how to loosen these restrictions.

To do that this pull request:
* Use the [is_protected_meta function](https://developer.wordpress.org/reference/functions/is_protected_meta/) to check if it is protected.
* Access the `$meta_keys `through [get_registered_meta_keys function](https://developer.wordpress.org/reference/functions/get_registered_meta_keys/) to check if the specific key has the `show_in_rest` property enabled.

## Testing Instructions

**Test it doesn't show the protected value when `show_in_rest` is false**


1. Register a custom field with `show_in_rest` set to false:

```PHP
register_meta(
	'post',
	'protected',
	array(
		'show_in_rest'   => false,
		'single'         => true,
		'type'           => 'string',
		'default'        => 'Protected value',
	)
);
```

3. Add a paragraph block in a page pointing to the protected block:

```html
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"protected"}}}}} -->
<p>Text</p>
<!-- /wp:paragraph -->
```

4. Save the page, go to the front and check it doesn't show the protected value.

**Test protected custom field**

1. Change the register source to `show_in_rest` true but protect it. It can be done adding a `_` at the beginning of the key or using a filter like this one:

```php
function protect_meta( $protected, $meta_key, $meta_type ) {
        return true;
}
add_filter( 'is_protected_meta', 'protect_meta', 10, 3 );
```

2. Check that the paragraph bound to the protected field doesn't show the protected value in the frontend.

---

Trac ticket: https://core.trac.wordpress.org/ticket/60651

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

SantosGuillamot committed:
Check if the meta field is protected

SantosGuillamot committed:
Check if the meta field is available in the REST API

swissspidy commented on src/wp-includes/block-bindings/post-meta.php:
```diff
@@ -34,6 +34,19 @@ function _block_bindings_post_meta_get_value( array $source_args, $block_instanc
+	$meta_keys = get_registered_meta_keys( 'post', $block_instance->context['postType'] );
+	// Add fields registered for all subtypes.
+	$meta_keys = array_merge( $meta_keys, get_registered_meta_keys( 'post', '' ) );
+	if ( empty( $meta_keys[ $source_attrs['key'] ]['show_in_rest'] ) || false === $meta_keys[ $source_attrs['key'] ]['show_in_rest'] ) {
```
`empty` handles `false` already

```suggestion
	if ( empty( $meta_keys[ $source_attrs['key'] ]['show_in_rest'] ) ) {
```

SantosGuillamot committed:
Remove unnecessary `show_in_rest` conditional

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

gziolo reviewed (approved):
That looks good, as discussed in the original PR in the Gutenberg repository. The approach taken might be to defensive, but we prefer to stay on the safe side and open it later if that makes more sense rather than take any risk something goes wrong.

gziolo commented:
As a follow-up during the RC phase, we should figure out how to cover the meta source with unit tests that prevent future regressions.

swissspidy commented:
I would appreciate unit tests as part of this change

gziolo commented:
I saw the comment on Trac and I agree with the reasoning. We will work on tests as soon as possible.

SantosGuillamot committed:
Use `source_args` instead of `source_attrs`

SantosGuillamot committed:
Add post meta bindings source tests

SantosGuillamot commented:
Adding tests totally makes sense to me. Thanks for pointing that out.

I've added some tests for the post meta source, covering the protected values, in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/1736c839baa0e5cd40f077e8ad93474030733fa7). I will take a deeper look tomorrow. Let me know any feedback 🙂

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,131 @@
+			)
+		);
+
+		$content = apply_filters( 'the_content', '<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"custom_field"}}}}} --><p>Fallback value</p><!-- /wp:paragraph -->' );
```
This is what I see in other places:

```php
$parsed_blocks = parse_blocks( '<!-- wp:tests/context-consumer /-->' );
render_block( $parsed_blocks[0] );
```

I assume that using `the_content` filter also works but it's a bit magical when you see it for the first time.

SantosGuillamot replied:
I tried using that approach first, but we need the post context to pass [this conditional](https://github.com/SantosGuillamot/wordpress-develop/blob/1736c839baa0e5cd40f077e8ad93474030733fa7/src/wp-includes/block-bindings/post-meta.php#L26-L28). If I just render a parsed block, there is no post ID.

gziolo replied:
I see, so it somehow sets the post ID in the context during rendering this way.

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,131 @@
+		// Removes custom fields registered by test cases.
+		$meta_keys = get_registered_meta_keys( 'post', '' );
+		foreach ( $meta_keys as $meta_key_name => $meta_key_value ) {
+			unregister_meta_key( 'post', $meta_key_name );
```
Is it safe to unregister all meta keys for the `post` type?

SantosGuillamot replied:
I added a check to unregister only the fields with "tests" in the name: [link](https://github.com/WordPress/wordpress-develop/pull/6197/commits/1fb2c47758c6deff8342b85b20d7ccaf15dd865c).

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,131 @@
+	public function test_protected_field_value_is_not_shown() {
+		register_meta(
+			'post',
+			'_protected_field',
```
`_` at the beginning makes it protected by design?

swissspidy replied:
Yes, by default. See `is_protected_meta()`

gziolo commented:
We also need to cover the following cases:
- post is password protected
- other reasons where the post can be viewable for the user

SantosGuillamot committed:
Unregister only the meta fields containing "tests"

SantosGuillamot committed:
Add test for password protected posts

SantosGuillamot committed:
Add test for posts that are not viewable

SantosGuillamot commented:
> We also need to cover the following cases:
> 
> * post is password protected
> * other reasons where the post can be viewable for the user

I wasn't sure how to handle these cases. I added two new tests for this using filters. Let me know if that's not correct.
* [Post is password protected](https://github.com/WordPress/wordpress-develop/pull/6197/commits/81d4e34b0ba73a9ace90cabf44b81296d1aff677)
* [Post is not viewable](https://github.com/WordPress/wordpress-develop/pull/6197/commits/8d0a0be3330450607c5b68f5cb19d8ee6b459d01)

xknown reviewed (commented):
It'd probably good to add other tests that do the following:
- Test for post meta keys that aren't registered (no register_meta call) to make sure any random post meta key is not exposed by default.
- Tests for meta values that might have unsafe HTML, are they sanitised?

peterwilsoncc commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,199 @@
+	/**
+	 * Tear down each test method.
+	 */
+	public function tear_down() {
```
This can be done via the global, see the meta rego tests

https://github.com/WordPress/wordpress-develop/blob/57b4193b9ae36b657cc23fec96f42af1add5abc2/tests/phpunit/tests/rest-api/rest-post-meta-fields.php#L11-L35

SantosGuillamot replied:
Thanks for the tip! 🙂  I've changed it in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/ce4aa0c78e953e4c2be1d285851f0585eda8e311).

peterwilsoncc commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,199 @@
+
+		parent::set_up();
+
+		$post = self::factory()->post->create_and_get( array() );
```
The post can be created in `wpSetUpBeforeClass` to avoid repeating hits to the DB.

SantosGuillamot replied:
Thanks for the tip! 🙂  I've changed it in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/ce4aa0c78e953e4c2be1d285851f0585eda8e311).

peterwilsoncc reviewed (commented):
A couple of notes for the tests.

I think this is a sensible change.

SantosGuillamot committed:
Move set up tp `wpSetUpBeforeClass`

SantosGuillamot committed:
Add test for non existing meta key

SantosGuillamot committed:
Remove filters in tests

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,218 @@
+			'The post content should show the fallback value instead of the custom field value.'
+		);
+
+		remove_filter(
```
Filters should be removed before the assertion, just after the usage. If the assertion fails, then the cleanup wouldn't happen.

SantosGuillamot replied:
Done as part of [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/dfe02a5b177792f4b881423ccb03f08a077d484f). I forgot to create separate ones 😄

SantosGuillamot committed:
Refactor the way post content is modified

SantosGuillamot commented:
In [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/7d253024900d285c4bef5117255bcc78724098b3), I've refactored the way the post content is updated because it seems not all tests were getting the correct post context.

I'd appreciate a new review on this one. I'm really sorry for all the changes, I just realized after trying to add more tests.

SantosGuillamot committed:
Add test to check unsafe html is sanitized

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+			)
+		);
+
+		$content = $this->getModifiedPostContent( '<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"tests_unsafe_html_field"}}}}} --><p>Fallback value</p><!-- /wp:paragraph -->' );
```
We probably will need more low-level tests for unsafe HTML because very likely it's the `the_content` filter that takes care of escaping the value here but on the block rendering level the HTML produced could be still unsafe.

gziolo replied:
Technically speaking, at the moment we should be good in terms of escaping the content from post meta:

https://github.com/WordPress/wordpress-develop/blob/9c415ed5fb9b4069208b9e43ff774acbfb3ef3a6/src/wp-includes/class-wp-block.php#L338

https://github.com/WordPress/wordpress-develop/blob/9c415ed5fb9b4069208b9e43ff774acbfb3ef3a6/src/wp-includes/class-wp-block.php#L373

What's I'd love to see is having a similar test on the block level to ensure future planned refactoring don't cause any regressions.

SantosGuillamot replied:
After thinking about this, I believe that this is not specific to the "Post meta" source, but all the sources. I created a similar tests in the block bindings render tests: [link](https://github.com/WordPress/wordpress-develop/pull/6197/commits/6414a972113b884220806f58613aa37be382e329).

I removed the one I added in the post meta source. I am not sure if it adds any value. Let me know if you want me to bring it back.

SantosGuillamot commented:
> It'd probably good to add other tests that do the following:
> 
> * Test for post meta keys that aren't registered (no register_meta call) to make sure any random post meta key is not exposed by default.
> * Tests for meta values that might have unsafe HTML, are they sanitised?

I've added a test for meta keys that aren't registered [here](https://github.com/WordPress/wordpress-develop/pull/6197/files#diff-b433ed8ede0ae232ea538e309b94dc6ff5d6debc4661ca993b791ad7f90b7c25R156-R169).

I've added a test to check unsafe HTML [here](https://github.com/WordPress/wordpress-develop/pull/6197/files#diff-b433ed8ede0ae232ea538e309b94dc6ff5d6debc4661ca993b791ad7f90b7c25R238-R262). _EDIT: I'll take a look at [this](https://github.com/WordPress/wordpress-develop/pull/6197#discussion_r1507330708)._

The block bindings API relies on the HTML API, which takes care of sanitizing the unsafe HTML.

Let me know if that is what you have in mind.

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+	/**
+	 * Update the global $post variable.
+	 */
+	private function updateGlobalPost() {
```
There isn't much value in having that code extracted from its method since it's used only once. It would make the code easier to follow if it was inlined.

You could also avoid adding inline comments by making the code structured like that:
```php
global $post;
$post = self::$post;
$post->post_content = $content;
return apply_filters( 'the_content', self::$post->post_content );
```

swissspidy replied:
The `self::$post` variable is useless anyway because the whole class only operates on the global post.

We can just always work on the global post if that is required.

SantosGuillamot replied:
I've tried using the global $post variable only for all the tests, but something is not working as I expected.

It seems that the value of the global `$post` variable is `undefined` (or the value I defined in `wpSetUpBeforeClass`) for the FIRST tests. For the following tests, the global `$post` variable becomes always null unless I redefine it.

This happens even if I remove all the logic and I just access that variable.

The way I'm solving that is storing the $post object in a private property when the post is initialized: [link](https://github.com/WordPress/wordpress-develop/pull/6197/files#diff-b433ed8ede0ae232ea538e309b94dc6ff5d6debc4661ca993b791ad7f90b7c25R30). And for each test I assign the global variable to that: [link](https://github.com/WordPress/wordpress-develop/pull/6197/files#diff-b433ed8ede0ae232ea538e309b94dc6ff5d6debc4661ca993b791ad7f90b7c25R50).

However, this feels like a workaround. Any idea if that is expected or how it could be solved?

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+	 *
+	 * @param string $content The new content.
+	 */
+	private function getModifiedPostContent( $content ) {
```
WordPress does not use camelCase. Please use snake_case everywhere (except for `wpSetUpBeforeClass` and `wpTearDownAfterClass`, these are special fixture methods). Consult the [coding standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/) for more details.

SantosGuillamot replied:
My bad, I assumed it worked that way only for tests. Changed in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/4cdf2419df23567913499d0d8dd82216dde49df8).

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+			)
+		);
+
+		function wp_tests_make_post_status_not_viewable() {
+			return false;
+		}
+
+		add_filter(
+			'is_post_status_viewable',
+			'wp_tests_make_post_status_not_viewable'
+		);
```
```suggestion
		add_filter( 'is_post_status_viewable', '__return_false' );
```

gziolo replied:
Right, these helper functions are handy 👍🏻

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/6e37a4c72e2dc5bd5271835dc340c95958653e78). I couldn't do it through GitHub interface 🤷

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+
+		$content = $this->getModifiedPostContent( '<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"tests_custom_field"}}}}} --><p>Fallback value</p><!-- /wp:paragraph -->' );
+
+		remove_filter(
+			'is_post_status_viewable',
+			'wp_tests_make_post_status_not_viewable'
+		);
```
```suggestion
		remove_filter( 'is_post_status_viewable', '__return_false' );
```

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/6e37a4c72e2dc5bd5271835dc340c95958653e78).

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+			)
+		);
+
+		function wp_tests_make_post_status_not_viewable() {
+			return false;
+		}
+
```
```suggestion
```

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/838927087f985c2da877f1b3f2ebc6d19cd8217b).

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+			)
+		);
+
+		function wp_tests_require_post_password() {
+			return true;
+		}
+
+		add_filter(
+			'post_password_required',
+			'wp_tests_require_post_password'
+		);
```
```suggestion
		add_filter( 'post_password_required', '__return_true' );
```

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/6e37a4c72e2dc5bd5271835dc340c95958653e78).

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+
+		$content = $this->getModifiedPostContent( '<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"tests_custom_field"}}}}} --><p>Fallback value</p><!-- /wp:paragraph -->' );
+
+		remove_filter(
+			'post_password_required',
+			'wp_tests_post_password_required'
+		);
```
```suggestion
		remove_filter( 'post_password_required', '__return_true' );
```

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/6e37a4c72e2dc5bd5271835dc340c95958653e78).

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+		self::$wp_meta_keys_saved = isset( $GLOBALS['wp_meta_keys'] ) ? $GLOBALS['wp_meta_keys'] : array();
+	}
+
+	/**
+	 * Tear down for every method.
+	 */
```
Incorrect. This is the tear down for the whole class.

```suggestion
	/**
	 * Tear down after class.
	 */
```

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	/**
+	 * Set up for every method.
```
Incorrect. This is for the whole class, only running once.

```suggestion
	 * Sets up shared fixtures.
```

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,263 @@
+	 * Tear down for every method.
+	 */
+	public static function wpTearDownAfterClass() {
+		$GLOBALS['wp_meta_keys'] = self::$wp_meta_keys_saved;
```
```suggestion
		$GLOBALS['wp_meta_keys'] = self::$wp_meta_keys_saved;
		unset( $GLOBALS['post'] );
```

SantosGuillamot committed:
Move unsafe html test to render bindings

gziolo commented:
> The block bindings API relies on the HTML API, which takes care of sanitizing the unsafe HTML.

That's only true for HTML attributes, as outlined in https://github.com/WordPress/wordpress-develop/pull/6197#discussion_r1507348594. For inner HTML, there is `wp_kses_post` call that takes care of it. It also made me realize that we need tests for both the attribute and inner HTML being connected to post meta.

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,237 @@
+	 * Set up for every method.
+	 */
+	public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
+		self::$post               = $factory->post->create_and_get();
```
As per my other comment, the static `$post` variable is redundant if you make it global anyway.

gziolo commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -234,30 +234,4 @@ public function test_custom_field_not_exposed_in_rest_api_is_not_shown() {
-	 *
-	 * @ticket 60651
-	 */
-	public function test_custom_field_with_unsafe_html_is_sanitized() {
```
I think it's fine to keep that test, too. It's better to stay on the safe side.

SantosGuillamot replied:
I added it back in [this commit](https://github.com/WordPress/wordpress-develop/pull/6197/commits/193131fc0c3e6c69cfc4b3646f61b5e9878e51df).

SantosGuillamot committed:
Add post meta unsafe html test again

SantosGuillamot committed:
Change how we modify the global $post variable

SantosGuillamot committed:
Use snake_case

SantosGuillamot committed:
Use hooks helper function
>
>
Co-authored-by: Pascal Birchler <swisspidy@git.wordpress.org>

SantosGuillamot committed:
Remove outdated function

SantosGuillamot committed:
Update class methods descriptions

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

SantosGuillamot committed:
Unset global post variable after class

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

SantosGuillamot commented:
> It also made me realize that we need tests for both the attribute and inner HTML being connected to post meta.

I assumed that if the attributes or the HTML are replaced properly should be covered for all sources in the binding render tests, right?

Do you think that having just one test specific for the post meta like [this one](https://github.com/WordPress/wordpress-develop/pull/6197/files#diff-b433ed8ede0ae232ea538e309b94dc6ff5d6debc4661ca993b791ad7f90b7c25R53-R76) would be enough or should we replicate all the tests?

gziolo commented on tests/phpunit/tests/block-bindings/render.php:
```diff
@@ -198,4 +198,40 @@ public function test_update_block_with_value_from_source_image_placeholder() {
+	 *
+	 * @covers ::register_block_bindings_source
+	 */
+	public function test_source_value_with_unsafe_html_is_sanitized() {
```
We should also add another test for a Button block where we put something tricky as the link's URL. Example: `javascript:alert('Unsafe HTML')`. I'm not sure if that is the best one, but that's definitely how someone could run JS by updating the dynamic source.

SantosGuillamot replied:
I'm wondering what if we should handle that case in the bindings because passing `javascript:alert('1')` to the href it is actually valid HTML. We would only be consuming the value stored in the custom field and exposed in the REST API. Not sure if it should be checked in the block bindings API, as part of the meta fields registration, or as part of the button block.

gziolo commented:
> Do you think that having just one test specific for the post meta like [this one](https://github.com/WordPress/wordpress-develop/pull/6197/files#diff-b433ed8ede0ae232ea538e309b94dc6ff5d6debc4661ca993b791ad7f90b7c25R53-R76) would be enough or should we replicate all the tests?

I hope it's enough because block rending covers the same functionality.

SantosGuillamot committed:
Add tests for bindings modifying html attribute

SantosGuillamot commented:
I added a new test for bindings modifying an HTML attribute like the image src.

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,270 @@
+	 */
+	public function set_up() {
+		parent::set_up();
+		// This seems to be needed to ensure that $GLOBALS['post'] is not null in all the tests.
```
Yes, turns out `tear_down()` actually resets those between tests.

```suggestion
		// Needed because tear_down() will reset it between tests.
```

swissspidy commented on tests/phpunit/tests/block-bindings/postMetaSource.php:
```diff
@@ -0,0 +1,270 @@
+	 */
+	public static function wpTearDownAfterClass() {
+		$GLOBALS['wp_meta_keys'] = self::$wp_meta_keys_saved;
+		unset( $GLOBALS['post'] );
```
Not needed because `tear_down()` does that already

```suggestion
```

swissspidy committed:
Apply suggestions from code review

swissspidy commented:
Committed in https://core.trac.wordpress.org/changeset/57754
